### PR TITLE
Adaptive VRR support

### PR DESCRIPTION
**Pull request by Drakulix**
Draft until cosmic-randr and cosmic-settings are ready

### CODE REVIEW

**Comment by Drakulix** [comment]
Now only blocked on smithay.

File: src/backend/kms/surface/mod.rs [hunk 1]
```diff
@@ -401,6 +405,33 @@ impl Surface {
...
+        let (tx, rx) = std::sync::mpsc::sync_channel(1);
+        let _ = self
+            .thread_command
+            .send(ThreadCommand::AdaptiveSyncAvailable(tx));
+        rx.recv().context("Surface thread died")?
+    }
+
+    pub fn use_adaptive_sync(&mut self, vrr: AdaptiveSync) -> Result<bool> {
+        if vrr != AdaptiveSync::Disabled {
+            let (tx, rx) = std::sync::mpsc::sync_channel(1);
```
Comment (ids1024) [inline src/backend/kms/surface/mod.rs:418]: I wish `std` had a oneshot channel, but I guess it's not worth adding a dependency for it when it's not already in the dependency tree.

File: src/config/mod.rs [hunk 2]
```diff
@@ -98,19 +98,34 @@ pub enum OutputState {
...
 
+#[derive(Debug, Deserialize, Serialize, Clone, Copy, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum AdaptiveSync {
+    #[serde(rename = "true")]
+    Enabled,
+    #[serde(rename = "false")]
+    Disabled,
+    Force,
+}
```
Comment (ids1024) [inline src/config/mod.rs:113]: It seems a little odd to use a different terminology (enabled/force vs automatic/always) from the protocol, though it doesn't really matter that much.

**Review by ids1024 — APPROVED** [review]
I'm not sure how best to test that VRR is actually working, but the changes here look good.

**Pull request merged**